File: HabitTracker.cs
// HabitTracker.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HabitTracker
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--add": opts.Add = args[++i]; break;
                    case "--remove": opts.Remove = args[++i]; break;
                    case "--check": opts.Check = args[++i]; break;
                    case "--list": opts.List = true; break;
                    case "--reset": opts.Reset = true; break;
                    case "--file": opts.File = args[++i]; break;
                    default: Console.WriteLine("Неизвестный аргумент"); return;
                }
            }
            var tracker = new Tracker(opts.File);
            if (!string.IsNullOrEmpty(opts.Add)) tracker.AddHabit(opts.Add);
            else if (!string.IsNullOrEmpty(opts.Remove)) tracker.RemoveHabit(opts.Remove);
            else if (!string.IsNullOrEmpty(opts.Check)) tracker.CheckHabit(opts.Check);
            else if (opts.List) tracker.ListHabits();
            else if (opts.Reset) tracker.ResetData();
            else Console.WriteLine("Используйте --help для справки");
        }

        class Options
        {
            public string Add { get; set; }
            public string Remove { get; set; }
            public string Check { get; set; }
            public bool List { get; set; }
            public bool Reset { get; set; }
            public string File { get; set; } = "habits.json";
        }

        class Tracker
        {
            private Dictionary<string, List<string>> habits;
            private readonly string filepath;

            public Tracker(string filepath)
            {
                this.filepath = filepath;
                Load();
            }

            private void Load()
            {
                if (File.Exists(filepath))
                {
                    string json = File.ReadAllText(filepath);
                    habits = JsonSerializer.Deserialize<Dictionary<string, List<string>>>(json) ?? new Dictionary<string, List<string>>();
                }
                else
                {
                    habits = new Dictionary<string, List<string>>();
                }
            }

            private void Save()
            {
                string json = JsonSerializer.Serialize(habits, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(filepath, json);
            }

            public void AddHabit(string name)
            {
                if (habits.ContainsKey(name))
                {
                    Console.WriteLine($"Привычка '{name}' уже существует.");
                    return;
                }
                habits[name] = new List<string>();
                Save();
                Console.WriteLine($"Привычка '{name}' добавлена.");
            }

            public void RemoveHabit(string name)
            {
                if (!habits.ContainsKey(name))
                {
                    Console.WriteLine($"Привычка '{name}' не найдена.");
                    return;
                }
                habits.Remove(name);
                Save();
                Console.WriteLine($"Привычка '{name}' удалена.");
            }

            public void CheckHabit(string name)
            {
                if (!habits.ContainsKey(name))
                {
                    Console.WriteLine($"Привычка '{name}' не найдена.");
                    return;
                }
                string today = DateTime.UtcNow.ToString("yyyy-MM-dd");
                if (habits[name].Contains(today))
                {
                    Console.WriteLine($"Привычка '{name}' уже отмечена сегодня.");
                    return;
                }
                habits[name].Add(today);
                Save();
                Console.WriteLine($"Привычка '{name}' отмечена на сегодня.");
            }

            public void ListHabits()
            {
                if (habits.Count == 0)
                {
                    Console.WriteLine("Нет привычек.");
                    return;
                }
                var today = DateTime.UtcNow;
                foreach (var kv in habits)
                {
                    var week = new List<string>();
                    for (int i = 6; i >= 0; i--)
                    {
                        week.Add(today.AddDays(-i).ToString("yyyy-MM-dd"));
                    }
                    int count = week.Count(d => kv.Value.Contains(d));
                    string bar = string.Join("", week.Select(d => kv.Value.Contains(d) ? "[x]" : "[ ]"));
                    Console.WriteLine($"Привычка: {kv.Key}");
                    Console.WriteLine($"  {bar}  ({count}/7, {count*100/7}%)");
                    Console.WriteLine($"  Всего дней: {kv.Value.Count}");
                }
            }

            public void ResetData()
            {
                habits.Clear();
                Save();
                Console.WriteLine("Все данные сброшены.");
            }
        }
    }
}

File: habit_tracker.rs
// habit_tracker.rs
use chrono::{Local, Duration};
use clap::{App, Arg};
use serde::{Deserialize, Serialize};
use serde_json;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

#[derive(Serialize, Deserialize)]
struct Data {
    habits: HashMap<String, Vec<String>>,
}

impl Default for Data {
    fn default() -> Self {
        Data { habits: HashMap::new() }
    }
}

struct Tracker {
    filepath: String,
    data: Data,
}

impl Tracker {
    fn new(filepath: &str) -> Self {
        let data = if Path::new(filepath).exists() {
            let content = fs::read_to_string(filepath).unwrap_or_default();
            serde_json::from_str(&content).unwrap_or_default()
        } else {
            Data::default()
        };
        Tracker { filepath: filepath.to_string(), data }
    }

    fn save(&self) {
        let json = serde_json::to_string_pretty(&self.data).unwrap();
        fs::write(&self.filepath, json).unwrap();
    }

    fn add_habit(&mut self, name: &str) {
        if self.data.habits.contains_key(name) {
            println!("Привычка '{}' уже существует.", name);
            return;
        }
        self.data.habits.insert(name.to_string(), Vec::new());
        self.save();
        println!("Привычка '{}' добавлена.", name);
    }

    fn remove_habit(&mut self, name: &str) {
        if !self.data.habits.contains_key(name) {
            println!("Привычка '{}' не найдена.", name);
            return;
        }
        self.data.habits.remove(name);
        self.save();
        println!("Привычка '{}' удалена.", name);
    }

    fn check_habit(&mut self, name: &str) {
        let dates = self.data.habits.get_mut(name);
        if let Some(dates) = dates {
            let today = Local::now().format("%Y-%m-%d").to_string();
            if dates.contains(&today) {
                println!("Привычка '{}' уже отмечена сегодня.", name);
                return;
            }
            dates.push(today);
            self.save();
            println!("Привычка '{}' отмечена на сегодня.", name);
        } else {
            println!("Привычка '{}' не найдена.", name);
        }
    }

    fn list_habits(&self) {
        if self.data.habits.is_empty() {
            println!("Нет привычек.");
            return;
        }
        let today = Local::now();
        for (name, dates) in &self.data.habits {
            // Последние 7 дней
            let week_dates: Vec<String> = (0..7)
                .map(|i| (today - Duration::days(6 - i)).format("%Y-%m-%d").to_string())
                .collect();
            let count = week_dates.iter().filter(|d| dates.contains(d)).count();
            let bar: String = week_dates
                .iter()
                .map(|d| if dates.contains(d) { "[x]" } else { "[ ]" })
                .collect();
            println!("Привычка: {}", name);
            println!("  {}  ({}/7, {}%)", bar, count, count * 100 / 7);
            println!("  Всего дней: {}", dates.len());
        }
    }

    fn reset(&mut self) {
        self.data.habits.clear();
        self.save();
        println!("Все данные сброшены.");
    }
}

fn main() {
    let matches = App::new("Трекер привычек")
        .arg(Arg::with_name("add").long("add").takes_value(true).help("Добавить привычку"))
        .arg(Arg::with_name("remove").long("remove").takes_value(true).help("Удалить привычку"))
        .arg(Arg::with_name("check").long("check").takes_value(true).help("Отметить выполнение"))
        .arg(Arg::with_name("list").long("list").help("Показать все привычки"))
        .arg(Arg::with_name("reset").long("reset").help("Сбросить данные"))
        .arg(Arg::with_name("file").long("file").takes_value(true).default_value("habits.json"))
        .get_matches();

    let file = matches.value_of("file").unwrap();
    let mut tracker = Tracker::new(file);

    if let Some(name) = matches.value_of("add") {
        tracker.add_habit(name);
    } else if let Some(name) = matches.value_of("remove") {
        tracker.remove_habit(name);
    } else if let Some(name) = matches.value_of("check") {
        tracker.check_habit(name);
    } else if matches.is_present("list") {
        tracker.list_habits();
    } else if matches.is_present("reset") {
        tracker.reset();
    } else {
        println!("Используйте --help для справки");
    }
}
